src/grad/primitives/gate.cr

```diff
@@ -34,5 +34,5 @@ abstract class Num::Grad::Gate(T)
   abstract def backward(payload : Num::Grad::Payload(T)) : Array(T)
 
   # Caches the result of an operation on a context
-  abstract def cache(result : Num::Grad::Variable(T), *args)
+  abstract def cache(result : Num::Grad::Variable(T), *args : Num::Grad::Variable(T))
```

Comment:
Is the issue with this that child methods sometimes say `args` is an instance of `Num::Grad::Variable`, which isn't explicitly defined by the abstract base, or that `args` must now have a defined type?

`cache` methods can take anything that a gate needs to store information.  For example, the convolutional gate also stores tuples of integers which match up with padding and stride.

Reply:
Hey Chris, the issue is the compiler now complains that the implementations are too strict
```
In src/grad/gates_blas.cr:25:1

 25 | class Num::Grad::MatMulGate(T) < Num::Grad::Gate(T)
      ^
Error: abstract `def Num::Grad::Gate(T)#cache(result : Num::Grad::Variable(T), *args)` must be implemented by Num::Grad::MatMulGate(T)
```

So either we restrict the abstract def or we remove the type restriction on each `def cache` definition

![image](https://user-images.githubusercontent.com/782103/109480254-38efcf00-7a73-11eb-8d73-5cba6025abfc.png)